src/undo_log.rs

```diff
@@ -0,0 +1,225 @@
+//! Module which contains the snapshot/rollback functionality of the `ena` data structures.
+//!
+//! For most usecases this is just an internal implementation detail. However if many `ena`
+//! data structures are used snapshotted simultaneously it is possible to use
+//! `UnificationTableStorage`/`SnapshotVecStorage` instead and use a custom `UndoLogs<T>`
+//! type capable of recording the actions of all used data structures.
+//!
+//! Since the `*Storage` variants do not have an undo log `with_log` must be called with the
+//! unified log before any mutating actions.
+
+/// A trait which allows actions (`T`) to be pushed which which allows the action to be undone at a
+/// later time if needed
+pub trait UndoLogs<T> {
+    fn in_snapshot(&self) -> bool {
+        self.num_open_snapshots() > 0
+    }
+    fn num_open_snapshots(&self) -> usize {
+        0
+    }
+    fn push(&mut self, undo: T);
```

Comment:
Pushes a new "undo item" onto the undo log. This method is invoked when some action is taken (e.g., a variable is unified). It records the info needed to reverse that action should an enclosing snapshot be rolleod back.